[SPARK-18278] Minimal support for submitting to Kubernetes.

mccheah opened the pull request:
Phase 1 for Kubernetes support in Spark. Only supports static resource allocation in cluster mode.

Some changes were made to the original fundamental approach to https://github.com/foxish/spark/pull/7, in particular how the REST server is built. Now, the REST server uses the existing submission REST server infrastructure as its base, instead of creating new Jetty-based code from scratch. Additionally, uploading local dependencies now uses a separate field in order to deduplicate from specifying jars local to the docker image via "spark.jars" vs. uploading jars from the client machine. The appropraite APIs to expose these configuration knobs are open to discussion, especially since the user's main resource is currently being uploaded indiscriminately but one could foresee the user wanting to specify their main resource as a file local to the Docker image's disk.

Also the client arguments have changed to mostly use Spark properties. Like the YARN support, common configuration points should be able to be set by arguments to spark-submit, but translated to properties in SparkConf.

mccheah committed:
[SPARK-18278] Minimal support for submitting to Kubernetes.

mccheah committed:
Fix style

mccheah committed:
Make naming more consistent

mccheah committed:
Fix building assembly with Kubernetes.

mccheah committed:
Service account support, use constants from fabric8 library.

ash211 commented:
@mccheah is this PR the one meant to be the phase one implementation (without the API layer)?  If so let's update the description to make that clear what this is intended to become.

mccheah committed:
Some small changes

- Don't hold the raw secret bytes
- Add CPU limits and requests

mccheah commented on kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala:
```diff
@@ -0,0 +1,269 @@
+            .addToRequests("memory", executorMemoryQuantity)
+            .addToLimits("memory", executorMemoryLimitQuantity)
+            .addToRequests("cpu", executorCpuQuantity)
+            .addToLimits("cpu", executorCpuQuantity)
```
If the container goes over this CPU limit is it forcefully killed? Or is the CPU limit merely what's provided on the hardware, and then it's impossible for the container to actually ever go over the limit?

foxish replied:
Details are in https://github.com/kubernetes/kubernetes/blob/master/docs/design/resource-qos.md
The container wouldn't get killed, as CPU is classified as a compressible resource. It would get throttled.

foxish replied:
If request == limit, we are making its QoS == guaranteed. It's the requested amount that is considered when making scheduling decisions. Having a higher limit would let us burst to use more CPU if it is available when overcommit is enabled in the cluster.

ash211 replied:
hmm so if I have 2 cores on a cluster and this Spark service needs 1 core minimum but would happily accept the second core if it was available, how should I set up cpu requests and limits?  request=1 and limit=2 ?

mccheah replied:
I'm not sure if we're expecting CPU bursts here - executors use a fixed-size thread pool based on the provided core count for running tasks. Setting limit == requested seems sufficient, but I'm not sure if this accounts for other random threads that are in the executor process.

foxish replied:
I wonder if we should have the QoS class of these executor pods be configurable. When we think of clusters sharing pods running serving workloads alongside Spark, the batch jobs would be lower priority, and it would make sense to make the executors be "best-effort" (no limit/request). They would get killed before the serving jobs if CPU/Memory were lacking.

mccheah replied:
I think the semantics we have here are closer to what is expected by Spark users migrating from YARN and Mesos.

foxish replied:
Setting a slightly higher limit than request for the CPU (10-25% more) may be a good thing, but if as Matt said, we are sure that we don't expect CPU bursts, it is fine as it is for the MVP.  I think in future, in order to improve utilization, we want to allow pods having different QoS classes. We probably want something similar to YARN's queueing mechanism which we can implicitly translate into QoS class. We can discuss this further after phase 1 is complete however.

ash211 commented on core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:
```diff
@@ -229,6 +230,7 @@ object SparkSubmit extends CommandLineUtils {
+      case m if m.startsWith("kubernetes") => KUBERNETES
       case m if m.startsWith("local") => LOCAL
       case _ =>
         printErrorAndExit("Master must either be yarn or start with spark, mesos, local")
```
this error message needs updating

ash211 commented on core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:
```diff
@@ -229,6 +230,7 @@ object SparkSubmit extends CommandLineUtils {
         YARN
       case m if m.startsWith("spark") => STANDALONE
       case m if m.startsWith("mesos") => MESOS
+      case m if m.startsWith("kubernetes") => KUBERNETES
```
I think we agreed we wanted kubernetes URLs to be of the form: `k8s://http://hostname:port` and `k8s://https://hostname:port`

ash211 commented on kubernetes/core/pom.xml:
```diff
@@ -0,0 +1,101 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
all the kubernetes files need to be under `/resource-managers/kubernetes` now that upstream has put yarn and mesos into the resource-managers folder

ash211 commented on kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/Client.scala:
```diff
@@ -0,0 +1,362 @@
+                .endContainer()
+              .endSpec()
+            .done()
+          submitCompletedFuture.get(30, TimeUnit.SECONDS)
```
I got this exception when trying to run a job and the image failed to pull:

```
Exception in thread "OkHttp https://10.0.23.119:6443/api/v1/namespaces/spark-test1/pods?labelSelector=driver-launcher-selector%3Ddriver-launcher-1483785139234&resourceVersion=10329&watch=true WebSocket" java.lang.NullPointerException
	at org.spark_project.guava.base.Preconditions.checkNotNull(Preconditions.java:191)
	at org.spark_project.guava.util.concurrent.AbstractFuture.setException(AbstractFuture.java:201)
	at org.spark_project.guava.util.concurrent.SettableFuture.setException(SettableFuture.java:68)
	at org.apache.spark.deploy.kubernetes.Client$$anonfun$run$5$$anon$1.onClose(Client.scala:177)
	at io.fabric8.kubernetes.client.dsl.internal.WatchConnectionManager$1.onClose(WatchConnectionManager.java:255)
	at okhttp3.internal.ws.RealWebSocket.peerClose(RealWebSocket.java:197)
	at okhttp3.internal.ws.RealWebSocket.access$200(RealWebSocket.java:38)
	at okhttp3.internal.ws.RealWebSocket$1$2.execute(RealWebSocket.java:84)
	at okhttp3.internal.NamedRunnable.run(NamedRunnable.java:32)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Exception in thread "main" java.util.concurrent.TimeoutException: Timeout waiting for task.
	at org.spark_project.guava.util.concurrent.AbstractFuture$Sync.get(AbstractFuture.java:276)
	at org.spark_project.guava.util.concurrent.AbstractFuture.get(AbstractFuture.java:96)
	at org.apache.spark.deploy.kubernetes.Client$$anonfun$run$5.org$apache$spark$deploy$kubernetes$Client$$anonfun$$createDriverPod$1(Client.scala:218)
	at org.apache.spark.deploy.kubernetes.Client$$anonfun$run$5$$anonfun$apply$2.apply(Client.scala:224)
	at org.apache.spark.deploy.kubernetes.Client$$anonfun$run$5$$anonfun$apply$2.apply(Client.scala:224)
	at org.apache.spark.util.Utils$.tryWithResource(Utils.scala:2508)
	at org.apache.spark.deploy.kubernetes.Client$$anonfun$run$5.apply(Client.scala:224)
	at org.apache.spark.deploy.kubernetes.Client$$anonfun$run$5.apply(Client.scala:102)
	at org.apache.spark.util.Utils$.tryWithResource(Utils.scala:2508)
	at org.apache.spark.deploy.kubernetes.Client.run(Client.scala:102)
	at org.apache.spark.deploy.kubernetes.Client$.main(Client.scala:361)
	at org.apache.spark.deploy.kubernetes.Client.main(Client.scala)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.spark.deploy.SparkSubmit$.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:753)
	at org.apache.spark.deploy.SparkSubmit$.doRunMain$1(SparkSubmit.scala:178)
	at org.apache.spark.deploy.SparkSubmit$.submit(SparkSubmit.scala:203)
	at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:117)
	at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)
```

```
aash@aash01-mac ~$ kubectl get pods -n spark-test1
NAME       READY     STATUS         RESTARTS   AGE
spark-pi   0/1       ErrImagePull   0          1m
aash@aash01-mac ~$
```

We should provide a better message instead of timing out here, something like `pod spark-pi in namespace spark-test1 never reached Running status.  Last observed status: ErrImagePull`

ash211 commented on kubernetes/docker-minimal-bundle/src/main/docker/driver/Dockerfile:
```diff
@@ -0,0 +1,26 @@
+FROM ubuntu:trusty
+
+# Upgrade package index
+# install a few other useful packages plus Open Jdk 7
```
prefer Java 8 over Java 7 in these templates

ash211 commented on kubernetes/docker-minimal-bundle/pom.xml:
```diff
@@ -0,0 +1,137 @@
+    <groupId>org.apache.spark</groupId>
+    <artifactId>spark-parent_2.11</artifactId>
+    <version>2.1.0-SNAPSHOT</version>
+    <relativePath>../pom.xml</relativePath>
```
`../../pom.xml`

foxish commented:
@mccheah I'm trying to build this and get compiler errors 
```
[INFO] Spark Project Shuffle Streaming Service ............ FAILURE [  0.874 s] 
```
My build command is 
```
./build/mvn -U -T 4 -Pkubernetes -Dhadoop.version=2.4.0 -Dmaven.test.skip=true -DskipTests package
```
Do you use a different one?

mccheah commented:
@foxish this works for me: `build/mvn compile package -DskipTests -Pkubernetes -Phadoop2.4`

mccheah committed:
Use k8s:// formatted URL instead of separate setting.

mccheah committed:
Merge remote-tracking branch 'apache/master' into k8s-support-alternate-incremental

foxish committed:
Reindent comment to conforn to JavaDoc style

The build process fails ScalaStyle checks otherwise.

ash211 commented on core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:
```diff
@@ -229,6 +230,7 @@ object SparkSubmit extends CommandLineUtils {
+      case m if m.startsWith("k8s") => KUBERNETES
       case m if m.startsWith("local") => LOCAL
       case _ =>
         printErrorAndExit("Master must either be yarn or start with spark, mesos, local")
```
please update this error message to have "k8s" in the list

mccheah committed:
Move kubernetes under resource-managers folder.

mccheah committed:
Merge remote-tracking branch 'apache-spark-on-k8s/k8s-support-alternate-incremental' into k8s-support-alternate-incremental

ash211 commented:
I think two of my earlier comments are still remaining:

- [ ] use java 8 instead of java 7 in the docker images
- [ ] Good error message for `ErrImagePull` error state

Can you please send those in as separate PRs into this one's branch?  It will be easier to review separately that way from the full PR

mccheah committed:
Use tar and gzip to compress+archive shipped jars (#2)

* Use tar and gzip to archive shipped jars.

* Address comments

* Move files to resolve merge

mccheah committed:
Use alpine and java 8 for docker images. (#10)

* Use alpine and java 8 for docker images.

* Remove installation of vim and redundant comment

mccheah committed:
Copy the Dockerfiles from docker-minimal-bundle into the distribution. (#12)

foxish committed:
inherit IO (#13)

ssuchter commented:
Hi folks -

In case anyone hasn't seen how to do this before, here's how you debug the failing Travis CI build. Click on one of the "Details" links - here's the one for the pull request:

https://travis-ci.org/apache-spark-on-k8s/spark/builds/191462919

then in that one, Travis has two separate CI builds, one for Java 7 and one for Java 8. Both failed, so I just clicked on the first one, #2.1. That gets me here:

https://travis-ci.org/apache-spark-on-k8s/spark/jobs/191462920

In the "Job log" on that page you can see the failure, it's in the Checkstyle linter:

[ERROR] src/main/java/org/apache/spark/launcher/SparkSubmitOptionParser.java:[83] (sizes) LineLength: Line is longer than 100 characters (found 102).

mccheah committed:
Error messages when the driver container fails to start. (#11)

* Error messages when the driver container fails to start.

* Fix messages a bit

* Use timeout constant

* Delete the pod if it fails for any reason (not just timeout)

* Actually set submit succeeded

* Fix typo

foxish committed:
Fix linter error to make CI happy (#18)

mccheah committed:
Documentation for the current state of the world (#16)

* Documentation for the current state of the world.

* Adding navigation links from other pages

* Address comments, add TODO for things that should be fixed

* Address comments, mostly making images section clearer

* Virtual runtime -> container runtime

mccheah committed:
Development workflow documentation for the current state of the world. (#20)

* Development workflow documentation for the current state of the world.

* Address comments.

* Clarified code change and added ticket link

foxish committed:
Added service name as prefix to executor pods (#14)

* Added service name as prefix to executor pods to be able to tell them apart from kubectl output

* Addressed comments

kimoonkim committed:
Add kubernetes profile to travis CI yml file (#21)

* Add kubernetes profile to travis yml file

* Fix long lines in CompressionUtils.scala

lins05 committed:
Improved the example commands in running-on-k8s document. (#25)

* Improved the example commands in running-on-k8s document.

* Fixed more example commands.

* Fixed typo.

foxish committed:
Fix spacing for command highlighting (#31)

mccheah committed:
Support custom labels on the driver pod. (#27)

* Support custom labels on the driver pod.

* Add integration test and fix logic.

* Fix tests

* Fix minor formatting mistake

* Reduce unnecessary diff

foxish committed:
Make pod name unique using the submission timestamp (#32)

mccheah committed:
A number of small tweaks to the MVP. (#23)

* A number of small tweaks to the MVP.

- Master protocol defaults to https if not specified
- Removed upload driver extra classpath functionality
- Added ability to specify main app resource with container:// URI
- Updated docs to reflect all of the above
- Add examples to Docker images, mostly for integration testing but
could be useful for easily getting started without shipping anything

* Add example to documentation.

ash211 committed:
Correct hadoop profile: hadoop2.7 -> hadoop-2.7 (#41)

See https://github.com/apache-spark-on-k8s/spark/blob/k8s-support-alternate-incremental/pom.xml#L2615

lins05 committed:
Support setting the driver pod launching timeout. (#36)

* Support setting the driver pod launching timeout.

And increase the default value from 30s to 60s. The current value of
30s is kind of short for pulling the image from public docker registry
plus the container/JVM start time.

* Use a better name for the default timeout.

ash211 committed:
Sanitize kubernetesAppId for use in secret, service, and pod names (#45)

kimoonkim committed:
Support spark.driver.extraJavaOptions (#48)

lins05 committed:
Use "extraScalaTestArgs" to pass extra options to scalatest. (#52)

* Use "extraTestArgLine" to pass extra options to scalatest.

Because the "argLine" option of scalatest is set in pom.xml and we can't
overwrite it from the command line.

Ref apache-spark-on-k8s/spark#37

* Added a default value for extraTestArgLine

* Use a better name.

* Added a tip for this in the dev docs.

mccheah committed:
Use OpenJDK8's official Alpine image. (#51)

mccheah committed:
Remove unused driver extra classpath upload code (#54)

lins05 committed:
Fix k8s integration tests (#44)

* Fixed k8s integration test

- Enable spark ui explicitly for in-process submit
- Fixed some broken assertions in integration tests
- Fixed a scalastyle error in SparkDockerImageBuilder.scala
- Log into target/integration-tests.log like other modules

* Fixed line length.

* CR

foxish committed:
Added GC to components (#56)

mccheah commented:
The build is still failing, but I checked the build run logs and the Scalastyle seems fine at the reported bad line. Does anyone have any ideas?

ash211 commented:
@mccheah I don't think `./dev/scalastyle` checks the kubernetes code right now -- I was debugging this a little last night but didn't finish it.

mccheah commented:
The line it's failing on is in `SparkSubmitOptionParser` which is in `core`. Line 83 is apparently longer than 100 characters but I checked and that line seems fine.

foxish commented:
I just hit re-run on the travis build in case the PR presubmit hasn't run in a while.

mccheah commented:
@foxish build still fails - I suspect the build is using an older checkout of the code since it's complaining about this line: https://github.com/apache-spark-on-k8s/spark/pull/1/files#diff-a061b6796fc188e8ad7964d3555c233aR83 - this line was removed in a patch along the way.

ash211 committed:
Create README to better describe project purpose (#50)

* Create README to better describe project purpose

* Add links to usage guide and dev docs

* Minor changes

mccheah committed:
Access the Driver Launcher Server over NodePort for app launch + submit jars (#30)

* Revamp ports and service setup for the driver.

- Expose the driver-submission service on NodePort and contact that as
opposed to going through the API server proxy
- Restrict the ports that are exposed on the service to only the driver
submission service when uploading content and then only the Spark UI
after the job has started

* Move service creation down and more thorough error handling

* Fix missed merge conflict

* Add braces

* Fix bad merge

* Address comments and refactor run() more.

Method nesting was getting confusing so pulled out the inner class and
removed the extra method indirection from createDriverPod()

* Remove unused method

* Support SSL configuration for the driver application submission (#49)

* Support SSL when setting up the driver.

The user can provide a keyStore to load onto the driver pod and the
driver pod will use that keyStore to set up SSL on its server.

* Clean up SSL secrets after finishing submission.

We don't need to persist these after the pod has them mounted and is
running already.

* Fix compilation error

* Revert image change

* Address comments

* Programmatically generate certificates for integration tests.

* Address comments

* Resolve merge conflicts

* Fix bad merge

* Remove unnecessary braces

* Fix compiler error

mccheah committed:
Extract constants and config into separate file. Launch => Submit. (#65)

* Extract constants and config into separate file. Launch => Submit.

* Address comments

* A small shorthand

* Refactor more ThreadUtils

* Fix scalastyle, use cached thread pool

* Tiny Scala style change

mccheah committed:
Retry the submit-application request to multiple nodes (#69)

* Retry the submit-application request to multiple nodes.

* Fix doc style comment

* Check node unschedulable, log retry failures

mccheah committed:
Allow adding arbitrary files (#71)

* Allow adding arbitrary files

* Address comments and add documentation

ash211 committed:
Fix NPE around unschedulable pod specs (#79)

ash211 committed:
Introduce blocking submit to kubernetes by default (#53)

* Introduce blocking submit to kubernetes by default

Two new configuration settings:
- spark.kubernetes.submit.waitAppCompletion
- spark.kubernetes.report.interval

* Minor touchups

* More succinct logging for pod state

* Fix import order

* Switch to watch-based logging

* Spaces in comma-joined volumes, labels, and containers

* Use CountDownLatch instead of SettableFuture

* Match parallel ConfigBuilder style

* Disable logging in fire-and-forget mode

Which is enabled with spark.kubernetes.submit.waitAppCompletion=false
(default: true)

* Additional log line for when application is launched

* Minor wording changes

* More logging

* Drop log to DEBUG

lins05 committed:
Do not wait for pod finishing in integration tests. (#84)

Since the example job are patched to never finish.

lins05 committed:
Check for user jars/files existence before creating the driver pod. (#86)

* Check for user jars/files existence before creating the driver pod.

Close apache-spark-on-k8s/spark#85

* CR

mccheah committed:
Use readiness probe instead of client-side ping. (#75)

* Use readiness probe instead of client-side ping.

Keep one ping() just as a sanity check, but otherwise set up the
readiness probe to report the container as ready only when the ping
endpoint can be reached.

Also add a liveliness probe for convenience and symmetry.

* Extract common HTTP get action

* Remove some code

* Add delay to liveliness check

* Fix merge conflicts.

* Fix more merge conflicts

* Fix more merge conflicts

* Revamp readiness check logic

* Add addresses ready condition to endpoints watch

* Rearrange the logic some more.

* Remove liveness probe, retry against servers

* Fix compiler error

* Fix another compiler error

* Delay between retries. Remove unintended test modification

* FIx another compiler error

* Extract method

* Address comments

* Deduplicate node addresses, use lower initial connect timeout

* Drop maxRetriesPerServer from 10 to 3

ash211 committed:
Note integration tests require Java 8 (#99)

foxish committed:
Bumping up kubernetes-client version to fix GKE and local proxy (#105)

* Bumping up kubernetes-client version to add fixes

* Modify wording

* Addressed comments

ash211 committed:
Truncate k8s hostnames to be no longer than 63 characters (#102)

* Truncate k8s hostnames to be no longer than 63 characters

* Use only executorId not executorKubernetesId

lins05 committed:
Fixed loading the executors page through the kubectl proxy. (#95)

Fix apache-spark-on-k8s/spark#87

foxish committed:
Filter nodes to only try and send files to external IPs (#106)

* Filter node addresses

* Added comment

ash211 committed:
Parse results of minikube status more rigorously (#97)

* Parse results of minikube status more rigorously

Prior code assumes the minikubeVM status line is always the first row output
from minikube status, and it is not when the version upgrade notifier prints
an upgrade suggestion message.

* Also filter ip response to expected rows

foxish committed:
Adding legacyHostIP to the list of IPs we look at (#114)

ash211 committed:
Add -DskipTests to dev docs (#115)

* Add -DskipTests to dev docs

* Remove extraneous skipTests

varunkatta committed:
Shutdown the thread scheduler in LoggingPodStatusWatcher on receiving job finish event notifications (#121)

kimoonkim committed:
Trigger scalatest plugin in the integration-test phase (#93)

* Trigger scalatest plugin in the integration-test phase

* Clean up unnecessary config section

foxish committed:
Fix issue with DNS resolution (#118)

* Fix issue with DNS resolution

* Address comments

mccheah committed:
Change the API contract for uploading local files (#107)

* Change the API contract for uploading local jars.

This mirrors similarly to what YARN and Mesos expects.

* Address comments

* Fix test

kimoonkim committed:
Optionally expose the driver UI port as NodePort (#131)

* Optionally expose driver UI as a NodePort

* Update the usage doc

* Rename serviceType -> uiServiceType

ash211 committed:
Set the REST service's exit code to the exit code of its driver subprocess (#138)

mccheah committed:
Pass the actual iterable from the option to get files (#139)

* Pass the actual iterable from the option to get files

* Split the original instance variables

* Explicitly set the type of the array

mccheah committed:
Use a separate class to track components that need to be cleaned up (#122)

* Refactor the cleaning up of Kubernetes components.

Create a KubernetesComponentsCleaner which can register arbitrary pods,
services, secrets, and ingresses. When an exception is thrown or the JVM
shuts down, the cleaner automatically purges any of its registered
components from Kubernetes. The components can be unregistered when the
driver successfully begins running, so that the application persists
beyond the lifetime of the spark-submit process.

* Fix spacing

* Address comments

* Fix compiler error

* Pull KubernetesComponentCleaner into instance variable

* Remove a parameter

* Remove redundant registerOrUpdateSecret for SSL

* Remove Ingresses from component cleaner

* Clear resources generically as opposed to specifying each type

* Remove incorrect test assertion

* Rename variable

kimoonkim committed:
Enable unit tests in Travis CI build (#132)

* Configure unit test build while banning flaky tests

* Clean up comment

ash211 committed:
Change driver pod's restart policy from OnFailure to Never (#145)

mccheah committed:
Extract SSL configuration handling to a separate class (#123)

* Extract SSL configuration to a separate class

* KubernetesSsl -> Ssl, container -> local

kimoonkim committed:
Exclude known flaky tests (#156)

* Exclude SparkSubmitSuite from Travis unit test build

* Remove SortShuffleSuite

* Exclude Java tests

foxish committed:
Richer logging and better error handling in driver pod watch (#154)

* pod-watch progress around watch events

* Simplify return

* comments

ash211 committed:
Document blocking submit calls (#152)

* Document blocking submit calls

https://github.com/apache-spark-on-k8s/spark/pull/53 added these config but didn't document them

* Update running-on-kubernetes.md

mccheah committed:
Allow custom annotations on the driver pod. (#163)

foxish committed:
Update client version & minikube version (#142)

* Update client version

* Upgrade minikube

* Update pom.xml

mccheah committed:
Allow customizing external URI provision + External URI can be set via annotations (#147)

* Listen for annotations that provide external URIs.

* FIx scalstyle

* Address comments

* Fix doc style

* Docs updates

* Clearly explain path rewrites

foxish committed:
Remove okhttp from top-level pom (#166)

mccheah committed:
Allow setting memory on the driver submission server. (#161)

* Allow setting memory on the driver submission server.

* Address comments

* Address comments

foxish committed:
Add a section for prerequisites (#171)

* Adding prerequisites

* address comments

foxish committed:
Add instructions to find master URL (#169)

mccheah committed:
Propagate exceptions (#172)

ash211 committed:
Logging for resource deletion (#170)

* Logging for resource deletion

Remove dangling colon and replace with an ellipses and a second log statement

* Update KubernetesResourceCleaner.scala

foxish committed:
Docs improvements (#176)

* Adding official alpha docker image to docs

* Reorder sections and create a specific one for "advanced"

* Provide limitations and instructions about running on GKE

* Fix title of advanced section: submission

* Improved section on running in the cloud

* Update versioning

* Address comments

* Address comments

ash211 committed:
Add Apache license to a few files (#175)

* Add Apache license to a few files

* Ignore license check on META-INF service

foxish committed:
Adding clarification pre-alpha (#181)

mccheah committed:
Allow providing an OAuth token for authenticating against k8s (#180)

* Allow providing an OAuth token for authenticating against k8s

* Organize imports

* Fix style

* Remove extra newline

* Use OAuth token data instead of a file.

mccheah commented:
Closing this in favor of diffing from branch-2.1-kubernetes. Probably will need to set the base branch to 2.1 also.

mccheah commented:
See #200
